File: 3633-earliest-finish-time-for-land-and-water-rides-i/3633-earliest-finish-time-for-land-and-water-rides-i.go
// int 타입의 최솟값을 구하는 헬퍼 함수
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// int 타입의 최댓값을 구하는 헬퍼 함수
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func earliestFinishTime(landStartTime []int, landDuration []int, waterStartTime []int, waterDuration []int) int {
	ans := math.MaxInt32

	// 플랜 1: 육상(Land) 놀이기구를 먼저 타고, 수상(Water) 놀이기구를 타는 경우
	for i := 0; i < len(landStartTime); i++ {
		landFinish := landStartTime[i] + landDuration[i]
		for j := 0; j < len(waterStartTime); j++ {
			// 수상 놀이기구는 자신의 시작 시간과 육상 놀이기구가 끝난 시간 중 더 늦은 시간에 시작할 수 있습니다.
			waterStart := max(waterStartTime[j], landFinish)
			finish := waterStart + waterDuration[j]
			ans = min(ans, finish)
		}
	}

	// 플랜 2: 수상(Water) 놀이기구를 먼저 타고, 육상(Land) 놀이기구를 타는 경우
	for i := 0; i < len(waterStartTime); i++ {
		waterFinish := waterStartTime[i] + waterDuration[i]
		for j := 0; j < len(landStartTime); j++ {
			// 육상 놀이기구는 자신의 시작 시간과 수상 놀이기구가 끝난 시간 중 더 늦은 시간에 시작할 수 있습니다.
			landStart := max(landStartTime[j], waterFinish)
			finish := landStart + landDuration[j]
			ans = min(ans, finish)
		}
	}

	return ans
}

File: 3633-earliest-finish-time-for-land-and-water-rides-i/3633-earliest-finish-time-for-land-and-water-rides-i.ts
function earliestFinishTime(
    landStartTime: number[],
    landDuration: number[],
    waterStartTime: number[],
    waterDuration: number[]
): number {
    let minFinishTime = Infinity;
    const n = landStartTime.length;
    const m = waterStartTime.length;

    // 시나리오 1: Land ride -> Water ride
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            // land ride가 끝나는 시간 = (시작 가능 시간 + 소요 시간)
            const landEnd = Math.max(landStartTime[i], landStartTime[i]) + landDuration[i];
            // water ride는 land ride가 끝난 후 또는 워터 라이드가 오픈했을 때 탈 수 있음
            const totalEnd = Math.max(landEnd, waterStartTime[j]) + waterDuration[j];
            minFinishTime = Math.min(minFinishTime, totalEnd);
        }
    }

    // 시나리오 2: Water ride -> Land ride
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            // water ride가 끝나는 시간
            const waterEnd = Math.max(waterStartTime[j], waterStartTime[j]) + waterDuration[j];
            // land ride는 water ride가 끝난 후 또는 랜드 라이드가 오픈했을 때 탈 수 있음
            const totalEnd = Math.max(waterEnd, landStartTime[i]) + landDuration[i];
            minFinishTime = Math.min(minFinishTime, totalEnd);
        }
    }

    return minFinishTime;
}
